File: AtCoder/ABC/209/cpp/D-Collision.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

const ll MAX_N = 100'000;

ll N, Q;
vector<ll> G[MAX_N];
ll dist[MAX_N];

void bfs() {
  queue<ll> q;
  fill(dist, dist + N, -1);
  dist[0] = 0;
  q.push(0);
  while (q.size()) {
    auto u = q.front();
    q.pop();
    for (auto v : G[u]) {
      if (dist[v] == -1) {
        dist[v] = dist[u] + 1;
        q.push(v);
      }
    }
  }
}

int main() {
  cin >> N >> Q;
  for (ll i = 0; i < N - 1; ++i) {
    ll a, b;
    cin >> a >> b;
    --a, --b;
    G[a].push_back(b);
    G[b].push_back(a);
  }
  bfs();
  while (Q--) {
    ll c, d;
    cin >> c >> d;
    --c, --d;
    if ((dist[c] - dist[d]) % 2 == 0) {
      puts("Town");
    } else {
      puts("Road");
    }
  }
}

File: AtCoder/ABC/209/python/D-Collision.py
from collections import deque


N, Q = map(int, input().split())
G = [[] for _ in range(N)]
for _ in range(N - 1):
    a, b = map(int, input().split())
    a -= 1
    b -= 1
    G[a].append(b)
    G[b].append(a)


dist = [-1] * N


def bfs():
    global dist
    dist[0] = 0
    q = deque([0])
    while q:
        u = q.popleft()
        for v in G[u]:
            if dist[v] < 0:
                dist[v] = dist[u] + 1
                q.append(v)


bfs()
for _ in range(Q):
    c, d = map(int, input().split())
    c -= 1
    d -= 1
    if (dist[c] - dist[d]) % 2 == 0:
        print("Town")
    else:
        print("Road")
